### ISSUE TITLE
Framing and @container:@set for scoped contexts

### ISSUE THREAD

**Issue by azaroth42** [w3c/json-ld-framing#64]
Playground:  http://tinyurl.com/y4ry8o3m

With a very simple identity frame (e.g. give the top node's id and a context), the expected output would be to follow the context's definition for `@container: @set` and wrap single objects in an array.  This works as expected for the context when applied via compaction (see the compaction tab in the playground).

However, in both the javascript and python implementations, the `part` property does not generate an array from the frame, where it does from compaction. Note that it is not due to Production within Production, as changing the class of either to (e.g.) `Activity` results in the same structure.

This seems like a bug?

**Comment by gkellogg** [comment]
A simpler formation still shows the problem ([playground](https://json-ld.org/playground/#startTab=tab-framed&copyContext=true&json-ld=%7B%22%40context%22%3A%7B%22%40version%22%3A1.1%2C%22%40vocab%22%3A%22http%3A%2F%2Fexample.org%2Fvocab%23%22%7D%2C%22%40id%22%3A%22http%3A%2F%2Fexample.org%2F1%22%2C%22%40type%22%3A%22HumanMadeObject%22%2C%22produced_by%22%3A%7B%22%40type%22%3A%22Production%22%2C%22_label%22%3A%22Top%20Production%22%2C%22part%22%3A%7B%22%40type%22%3A%22Production%22%2C%22_label%22%3A%22Test%20Part%22%7D%7D%7D&frame=%7B%22%40context%22%3A%7B%22%40version%22%3A1.1%2C%22%40vocab%22%3A%22http%3A%2F%2Fexample.org%2Fvocab%23%22%2C%22Production%22%3A%7B%22%40context%22%3A%7B%22part%22%3A%7B%22%40type%22%3A%22%40id%22%2C%22%40container%22%3A%22%40set%22%7D%7D%7D%7D%2C%22%40id%22%3A%22http%3A%2F%2Fexample.org%2F1%22%7D)).

```json
{
  "@context": {
    "@version": 1.1,
    "@vocab": "http://example.org/vocab#"
  },
  "@id": "http://example.org/1",
  "@type": "HumanMadeObject",
  "produced_by": {
    "@type": "Production",
    "_label": "Top Production",
    "part": {
      "@type": "Production",
      "_label": "Test Part"
    }
  }
}
```

Frame:
```json
{
  "@context": {
    "@version": 1.1,
    "@vocab": "http://example.org/vocab#",
    "Production": {
      "@context": {
        "part": {
          "@type": "@id",
          "@container": "@set"
        }
      }
    }
  },
  "@id": "http://example.org/1"
}
```

The issue seems to be in the last paragraph of [4.2.2](https://www.w3.org/TR/json-ld11-framing/#framing-algorithm):

> Recursively, replace all entries in compacted results where the key is `@preserv`e with the value of the entry. If the value of the entry is `@null`, replace the value with `null`. `If, after replacement, an array contains a single array value, replace the array with that value.` If, after replacement, an array contains only the value `null` remove the value, leaving an empty array.

The key bit recently added is the _italicized_ bit, which does not consider the context.

Logically, it would seem to make sense to do this last bit before compaction, although I suspect there was a reason we didn't do this.

**Comment by gkellogg** [comment]
So, the problem with cleaning up `@preserve` and `@null` before compaction, is that, in some cases, the `null` value should be kept when framing, but eliminated if it's just one component of an array. I've never understood why this was important, but perhaps @dlongley could shed some light on this.

The simplest thing would be do just do that cleanup before compaction, which would allow the regular compaction dynamics to work.

If keeping the `null` output is important, we could separate replacing `@null` with `null` until after compaction, although this would still likely leave odd results, with some values of `null` removed, and others remaining.

Preserving `null` must have had some value, otherwise, the whole use for `@null` is questionable.

Otherwise, we may need to compact, do `@preserve/@null` cleanup, then expand and re-compact to get this right, which seems silly and wastefull.

**Comment by gkellogg** [comment]
With a trial implementation, 18 framing tests fail, where `null` is expected and an empty array is what is actually produced.

For example, in `#t0005` we expect the following:
```json
{
  "@context": {
    "dcterms": "http://purl.org/dc/terms/",
    "ex": "http://example.org/vocab#"
  },
  "@graph": [
    {
      "@id": "http://example.org/test/#library",
      "@type": "ex:Library",
      "ex:contains": {
        "@id": "http://example.org/test#book",
        "@type": "ex:Book",
        "dcterms:title": "My Book",
        "ex:contains": {
          "@id": "http://example.org/test#chapter",
          "@type": "ex:Chapter",
          "dcterms:title": "Chapter One",
          "ex:null": null
        }
      }
    }
  ]
}
```

But, what is actually produced is:
```json
{
  "@context": {
    "dcterms": "http://purl.org/dc/terms/",
    "ex": "http://example.org/vocab#"
  },
  "@graph": [
    {
      "@id": "http://example.org/test/#library",
      "@type": "ex:Library",
      "ex:contains": {
        "@id": "http://example.org/test#book",
        "@type": "ex:Book",
        "dcterms:title": "My Book",
        "ex:contains": {
          "@id": "http://example.org/test#chapter",
          "@type": "ex:Chapter",
          "dcterms:title": "Chapter One",
          "ex:null": []
        }
      }
    }
  ]
}
```

**Comment by iherman** [comment]
This issue was discussed in [a meeting](https://www.w3.org/2018/json-ld-wg/Meetings/Minutes/2019/2019-08-16-json-ld#section4-1).

- `RESOLVED:  merge api#134 and syntax#212, and close syntax#204`
<details><summary><i class="differentiate">View the transcript</i></summary>
Benjamin Young: See Framing issue #64<br/>
Rob Sanderson: In framing context, I had a `<code>@container:@set`.</code> I did various things in playground, and discovered that in framing, container:set does not get handled properly. It does work fine in compaction.<br/>
… Also in Python implementation.<br/>
Gregg Kellogg: This is because there’s a step to handle <code>@preserve</code> values post-compaction.<br/>
… This is because there are cases where there are null values that are replaced during compaction.<br/>
… My suggested fix would emit empty array instead of null. But this does not preserve null in output anymore.<br/>
Dave Longley: JSON devs expect null there, so it must be preserved. They want to see the null there.<br/>
Gregg Kellogg: Except if null is part of an array.<br/>
Dave Longley: Indeed. That’s another case.<br/>
… I wonder if we can do something during compaction to indicate preserve:null.<br/>
… Can we leave a note during compaction for this?<br/>
Gregg Kellogg: One way would be to separate processing of preserve:null, and do preserve before compaction. Once it has been compacted, there may be values with null. If in array, would be removed, if not, it would be preserved.<br/>
… There’s no easy way to handle this. But handling null after compaction may be the best way.<br/>
Dave Longley: Not great, but may be a good solution.<br/>
Gregg Kellogg: Not needed to discuss this further during this call. There is a way to fix this. But this is implementation work.<br/>
4.2. Link header for HTML and JSON-LD #204<br/>
Benjamin Young: Link header: preemptive conneg.<br/>
Benjamin Young: See Syntax issue #204<br/>
Benjamin Young: See API PR #134<br/>
Benjamin Young: See Syntax PR #212<br/>
Benjamin Young: 2 PRs are ready to go, right?<br/>
Gregg Kellogg: Yes. Open question for base during redirect.<br/>
… If you retrieve something non-json. Doc loader will detect rel-alternative links, and redirect to that resource.<br/>
… I think the original requested document should be preserved as base URL.<br/>
… Just like 303 redirect semantics. In 302, this would not be the case.<br/>
Rob Sanderson: JS will silently follow redirects and give you the results. So you don’t know which redirects got you there.<br/>
Gregg Kellogg: JS should correctly implement redirect semantics.<br/>
Dave Longley: +1 to using 303 semantics, so on, +1 to gregg’s position.<br/>
Gregg Kellogg: You have processing state to keep track on this.<br/>
Proposed resolution: merge api#134 and syntax#212, and close syntax#204 (Ivan Herman)<br/>
Rob Sanderson: +1<br/>
Dave Longley: +1<br/>
Benjamin Young: +1<br/>
Rob Sanderson: :shipit:<br/>
Ivan Herman: +1<br/>
Gregg Kellogg: +1<br/>
Ruben Taelman: +0.5<br/>
Resolution #2: merge api#134 and syntax#212, and close syntax#204<br/>
4.<br/>
</details>
